src/linear_algebra/lagrange.lean

```diff
@@ -1,224 +1,353 @@
-of degree less than `s.card`. -/
-def fun_equiv_degree_lt : degree_lt F s.card ≃ₗ[F] (s → F) :=
-{ to_fun := λ f x, f.1.eval x,
-  map_add' := λ f g, funext $ λ x, eval_add,
-  map_smul' := λ c f, funext $ by simp,
-  inv_fun := λ f, ⟨interpolate s (λ x, if hx : x ∈ s then f ⟨x, hx⟩ else 0),
-    mem_degree_lt.2 $ degree_interpolate_lt _ _⟩,
-  left_inv := λ f, begin apply subtype.eq,
-    simp only [subtype.coe_mk, subtype.val_eq_coe, dite_eq_ite],
-    convert eq_interpolate s f (mem_degree_lt.1 f.2) using 1,
-    rw interpolate_eq_of_eval_eq,
-    intros x hx,
-    rw if_pos hx end,
-  right_inv := λ f, funext $ λ ⟨x, hx⟩, begin
-    convert eval_interpolate s _ x hx,
-    simp_rw dif_pos hx end }
-
-theorem interpolate_eq_interpolate_erase_add {x y : F} (hx : x ∈ s) (hy : y ∈ s) (hxy : x ≠ y) :
-  interpolate s f =
-  C (y - x)⁻¹ * ((X - C x) * interpolate (s.erase x) f + (C y - X) * interpolate (s.erase y) f) :=
+theorem interpolate_eq_of_eval_eq (hs : ∀ i, r i = s i) :
+ interpolate v r = interpolate v s := by {rw ← @function.funext_iff _ _ r s at hs, rw hs}
+
+def linterpolate (v : ι ↪ F) : (ι → F) →ₗ[F] polynomial F :=
+{ to_fun := interpolate v,
+ map_add' := λ f g, by { simp_rw [interpolate_def, ← finset.sum_add_distrib, ← add_mul, ← C_add],
+  refl },
+ map_smul' := λ c f, by { simp_rw [interpolate_def, finset.smul_sum, C_mul', smul_smul], refl } }
+
+lemma interpolate_add (r s) :
+ interpolate v (r + s) = interpolate v r + interpolate v s := (linterpolate v).map_add r s
+
+lemma interpolate_zero : interpolate v 0 = 0 :=
+(linterpolate v).map_zero
+
+lemma interpolate_neg (r) : interpolate v (-r) = -interpolate v r :=
+(linterpolate v).map_neg r
+
+lemma interpolate_sub (r s) :
+ interpolate v (r - s) = interpolate v r - interpolate v s := (linterpolate v).map_sub r s
+
+lemma interpolate_smul (c : F) (r) :
+ interpolate v (c • r) = c • interpolate v r := (linterpolate v).map_smul c r
+
+theorem eq_interpolate (f : F[X]) (hf : f.degree < fintype.card ι) :
+ interpolate v (λ i, eval (v i) f) = f :=
+eq_of_eval_eq' v degree_interpolate_lt hf (λ _, by simp_rw eval_interpolate)
+
+/-- TODO NEW DOCSTRING -/
+def fun_equiv_degree_lt : degree_lt F (fintype.card ι) ≃ₗ[F] (ι → F) :=
+{ to_fun := λ f i, f.1.eval (v i),
+ map_add' := λ f g, funext $ λ v, eval_add,
+ map_smul' := λ c f, funext $ by simp,
+ inv_fun := λ r, ⟨interpolate v r, mem_degree_lt.mpr degree_interpolate_lt⟩,
+ left_inv := λ f, subtype.eq (eq_interpolate _ (mem_degree_lt.1 f.2)),
+ right_inv := λ f, funext (λ i, eval_interpolate) }
+
+end interpolate
+
+section interpolate_at
+variables {ι : Type*} [fintype ι] [decidable_eq ι] {v : ι ↪ F} {i j : ι} {f g : F → F}
+
+def interpolate_at (v : ι ↪ F) (f : F → F) := interpolate v (λ i, f (v i))
+
+lemma interpolate_at_def : interpolate_at v f = interpolate v (λ i, f (v i)) := rfl
+
+theorem interpolate_at_eq_of_eval_eq (hs : ∀ i, f (v i) = g (v i)) :
```

Comment:
Maybe this could be an `ext` lemma?

Reply:
I considered this - clearly it's doing that kind of thing. `interpolate_eq_of_eval_eq` should probably also be. Is this the standard form?